mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp

```diff
@@ -0,0 +1,964 @@
+        this->numberOfCacheLinesAccessed,
+        this->loopIterationCounts);
+    long double temporalCost = getTemporalReuseCost(perm);
+
+    // Assumption: costs due to parallelism (synchronization) are 100x
+    // more expensive than those due to locality.
+    long double cost =
+        100 * parallelCost + spatialCost + temporalCost;
+
+    if (cost < minCost) {
+      minCost = cost;
+      bestPerm = perm;
+    }
+  }
+
+  //`permuteLoops()` (called in next step) maps loop 'i' to location
+  // bestPerm[i]. But here bestPerm[i] maps to loop
+  // at depth 'i'. Hence, we need to prepare a custom `returnPerm`
+  // according to requirements of `permuteLoops()`.
+  SmallVector<unsigned, 4> returnPerm;
+  returnPerm.resize(bestPerm.size());
+  for (unsigned i = 0; i < bestPerm.size(); i++)
+    returnPerm[bestPerm[i]] = i;
+  return returnPerm;
+}
+
+/// Resets all the protected data members alongwith provided arguments.
+void LoopInterchange::clear(SmallVector<AffineForOp, 4> &loopNest,
+                            DenseMap<Value, unsigned> &indVarIdMap) {
+  loopNest.clear();
+  indVarIdMap.clear();
+  this->loopIterationCounts.clear();
+  this->loopCarriedDependenceVector.clear();
+  this->numberOfCacheLinesAccessed.clear();
+  this->referenceGroups.clear();
+  this->elementSize.clear();
+  this->loopAccessMatrices.clear();
+  this->constVector.clear();
+}
+
+void LoopInterchange::runOnFunction() {
+  SmallVector<AffineForOp, 4> loopNest;
+
+  // A map to hold IDs for all affine.for induction vars.
+  DenseMap<Value, unsigned> indVarIdMap;
+
+  Operation *function = getFunction().getOperation();
+
+  handleImperfectlyNestedAffineLoops(*function);
+
+  (*function).walk([&](AffineForOp op) {
+    loopNest.push_back(op);
+
+    // op is root of some loop nest.
+    if ((op.getParentOp()->getName().getStringRef().str() == "func")) {
+
+      // The loop nest should not have any affine.if op and should have
+      // rectangular shaped iteration space.
+      if (!hasAffineIfStatement(op) &&
+          isRectangularAffineForLoopNest(loopNest)) {
+
+        // With a postorder traversal, affine.forops in `loopNest`
+        // are pushed in reverse order. We need to reverse this order to
+        // arrange them in loop nest order.
+        std::reverse(loopNest.begin(), loopNest.end());
+
+        unsigned loopIVCount = 0;
+        for (auto &op : loopNest) {
+          // Assign IDs to loop variables - used later to populate accessMatrix
+          // of each memref access.
+          Value indVar = op.getInductionVar();
+          indVarIdMap[indVar] = loopIVCount;
+          loopIVCount++;
+          // Populate `loopIterationCounts` of each for loop in loop nest.
+          this->loopIterationCounts.push_back(
+              (op.getConstantUpperBound() - op.getConstantLowerBound()) /
+              op.getStep());
+        }
+
+        SmallVector<unsigned, 4> bestPermutation =
+            getBestPermutation(loopNest, indVarIdMap);
```

Comment:
Pass an output argument by reference to avoid copies.

Reply:
Done.